File: app/src/main/java/mega/privacy/android/app/modalbottomsheet/OfflineOptionsBottomSheetDialogFragment.java
package mega.privacy.android.app.modalbottomsheet;

import static mega.privacy.android.app.utils.Constants.HANDLE;
import static mega.privacy.android.app.utils.Constants.THUMB_MARGIN_DP;
import static mega.privacy.android.app.utils.Constants.THUMB_SIZE_DP;
import static mega.privacy.android.app.utils.FileUtil.getFileFolderInfo;
import static mega.privacy.android.app.utils.FileUtil.getFileInfo;
import static mega.privacy.android.app.utils.FileUtil.isFileAvailable;
import static mega.privacy.android.app.utils.Util.dp2px;
import static mega.privacy.android.app.utils.Util.isOnline;
import static mega.privacy.android.app.utils.Util.scaleWidthPx;

import android.annotation.SuppressLint;
import android.net.Uri;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.LinearLayout;
import android.widget.RelativeLayout;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.facebook.drawee.view.SimpleDraweeView;

import java.io.File;
import java.util.Collections;

import mega.privacy.android.app.MegaOffline;
import mega.privacy.android.app.MimeTypeList;
import mega.privacy.android.app.R;
import mega.privacy.android.app.main.ManagerActivity;
import mega.privacy.android.app.utils.OfflineUtils;
import timber.log.Timber;

public class OfflineOptionsBottomSheetDialogFragment extends BaseBottomSheetDialogFragment implements View.OnClickListener {

    private MegaOffline nodeOffline = null;

    private File file;

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        contentView = View.inflate(getContext(), R.layout.bottom_sheet_offline_item, null);
        itemsLayout = contentView.findViewById(R.id.items_layout);

        if (savedInstanceState != null) {
            String handle = savedInstanceState.getString(HANDLE);
            nodeOffline = dbH.findByHandle(handle);
        } else if (requireActivity() instanceof ManagerActivity) {
            nodeOffline = ((ManagerActivity) requireActivity()).getSelectedOfflineNode();
        }

        return contentView;
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        contentView.findViewById(R.id.option_download_layout).setOnClickListener(this);
        contentView.findViewById(R.id.option_properties_layout).setOnClickListener(this);
        TextView optionInfoText = contentView.findViewById(R.id.option_properties_text);

        SimpleDraweeView nodeThumb = contentView.findViewById(R.id.offline_thumbnail);
        TextView nodeName = contentView.findViewById(R.id.offline_name_text);
        TextView nodeInfo = contentView.findViewById(R.id.offline_info_text);
        LinearLayout optionOpenWith = contentView.findViewById(R.id.option_open_with_layout);
        LinearLayout optionShare = contentView.findViewById(R.id.option_share_layout);

        contentView.findViewById(R.id.option_delete_offline_layout).setOnClickListener(this);
        optionOpenWith.setOnClickListener(this);
        optionShare.setOnClickListener(this);

        View separatorOpen = contentView.findViewById(R.id.separator_open);

        nodeName.setMaxWidth(scaleWidthPx(200, getResources().getDisplayMetrics()));
        nodeInfo.setMaxWidth(scaleWidthPx(200, getResources().getDisplayMetrics()));

        if (nodeOffline != null) {
            optionInfoText.setText(R.string.general_info);

            if (nodeOffline.isFolder()) {
                optionOpenWith.setVisibility(View.GONE);
                separatorOpen.setVisibility(View.GONE);
            } else {
                optionOpenWith.setVisibility(View.VISIBLE);
                separatorOpen.setVisibility(View.VISIBLE);
            }

            nodeName.setText(nodeOffline.getName());

            Timber.d("Set node info");
            file = OfflineUtils.getOfflineFile(requireContext(), nodeOffline);
            if (!isFileAvailable(file)) return;

            if (file.isDirectory()) {
                nodeInfo.setText(getFileFolderInfo(file, requireContext()));
            } else {
                nodeInfo.setText(getFileInfo(file, requireContext()));
            }

            if (file.isFile()) {
                if (MimeTypeList.typeForName(nodeOffline.getName()).isImage()) {
                    if (file.exists()) {
                        RelativeLayout.LayoutParams params = (RelativeLayout.LayoutParams) nodeThumb.getLayoutParams();
                        params.height = params.width = dp2px(THUMB_SIZE_DP);
                        int margin = dp2px(THUMB_MARGIN_DP);
                        params.setMargins(margin, margin, margin, margin);
                        nodeThumb.setLayoutParams(params);

                        nodeThumb.setImageURI(Uri.fromFile(file));
                    } else {
                        nodeThumb.setActualImageResource(MimeTypeList.typeForName(nodeOffline.getName()).getIconResourceId());
                    }
                } else {
                    nodeThumb.setImageResource(MimeTypeList.typeForName(nodeOffline.getName()).getIconResourceId());
                }
            } else {
                nodeThumb.setImageResource(R.drawable.ic_folder_list);
            }

            if (nodeOffline.isFolder() && !isOnline(requireContext())) {
                optionShare.setVisibility(View.GONE);
                contentView.findViewById(R.id.separator_share).setVisibility(View.GONE);
            }
        }

        super.onViewCreated(view, savedInstanceState);
    }

    @SuppressLint("NonConstantResourceId")
    @Override
    public void onClick(View v) {
        int id = v.getId();
        if (id == R.id.option_delete_offline_layout) {
            if (requireActivity() instanceof ManagerActivity) {
                ((ManagerActivity) requireActivity())
                        .showConfirmationRemoveFromOffline(nodeOffline,
                                this::setStateBottomSheetBehaviorHidden);
            }
        } else if (id == R.id.option_open_with_layout) {
            OfflineUtils.openWithOffline(requireContext(), Long.valueOf(nodeOffline.getHandle()));
        } else if (id == R.id.option_share_layout) {
            OfflineUtils.shareOfflineNode(requireContext(), Long.valueOf(nodeOffline.getHandle()));
        } else if (id == R.id.option_download_layout) {
            ((ManagerActivity) requireActivity()).saveOfflineNodesToDevice(
                    Collections.singletonList(nodeOffline));
        } else if (id == R.id.option_properties_layout) {
            ((ManagerActivity) requireActivity()).showOfflineFileInfo(nodeOffline);
        }

        setStateBottomSheetBehaviorHidden();
    }

    @Override
    public void onSaveInstanceState(Bundle outState) {
        super.onSaveInstanceState(outState);
        String handle = nodeOffline.getHandle();
        outState.putString(HANDLE, handle);
    }
}

File: app/src/main/java/mega/privacy/android/app/modalbottomsheet/OfflineOptionsBottomSheetDialogFragment.kt
package mega.privacy.android.app.modalbottomsheet

import android.content.Intent
import android.net.Uri
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.RelativeLayout
import androidx.core.os.bundleOf
import androidx.core.view.isGone
import mega.privacy.android.app.MegaOffline
import mega.privacy.android.app.MimeTypeList.Companion.typeForName
import mega.privacy.android.core.R as CoreUiR
import mega.privacy.android.app.R
import mega.privacy.android.app.activities.OfflineFileInfoActivity
import mega.privacy.android.app.databinding.BottomSheetOfflineItemBinding
import mega.privacy.android.app.presentation.offline.adapter.OfflineNodeListener
import mega.privacy.android.app.main.ManagerActivity
import mega.privacy.android.app.presentation.extensions.parcelable
import mega.privacy.android.app.utils.Constants
import mega.privacy.android.app.utils.FileUtil
import mega.privacy.android.app.utils.OfflineUtils
import mega.privacy.android.app.utils.Util
import timber.log.Timber

internal class OfflineOptionsBottomSheetDialogFragment : BaseBottomSheetDialogFragment(),
    View.OnClickListener {
    private val nodeOffline: MegaOffline by lazy(LazyThreadSafetyMode.NONE) {
        requireNotNull(requireArguments().parcelable(EXTRA_OFFLINE))
    }
    private var _binding: BottomSheetOfflineItemBinding? = null
    val binding: BottomSheetOfflineItemBinding
        get() = _binding!!

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?,
    ): View {
        contentView = BottomSheetOfflineItemBinding.inflate(inflater, container, false).also {
            _binding = it
            itemsLayout = it.itemsLayout
        }.root
        return contentView
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        binding.optionDownloadLayout.setOnClickListener(this)
        binding.optionPropertiesLayout.setOnClickListener(this)
        binding.optionShareLayout.setOnClickListener(this)
        binding.optionDeleteOfflineLayout.setOnClickListener(this)
        binding.optionOpenWithLayout.setOnClickListener(this)
        binding.offlineNameText.maxWidth = Util.scaleWidthPx(200, resources.displayMetrics)
        binding.offlineInfoText.maxWidth = Util.scaleWidthPx(200, resources.displayMetrics)
        binding.optionPropertiesText.setText(R.string.general_info)
        binding.optionOpenWithLayout.isGone = nodeOffline.isFolder
        binding.separatorOpen.isGone = nodeOffline.isFolder
        binding.offlineNameText.text = nodeOffline.name
        Timber.d("Set node info")
        val file = OfflineUtils.getOfflineFile(requireContext(), nodeOffline)
        if (!FileUtil.isFileAvailable(file)) return
        if (file.isDirectory) {
            binding.offlineInfoText.text =
                FileUtil.getFileFolderInfo(file, requireContext())
        } else {
            binding.offlineInfoText.text = FileUtil.getFileInfo(file, requireContext())
        }
        if (file.isFile) {
            if (typeForName(nodeOffline.name).isImage) {
                if (file.exists()) {
                    val params =
                        binding.offlineThumbnail.layoutParams as RelativeLayout.LayoutParams
                    params.width = Util.dp2px(Constants.THUMB_SIZE_DP.toFloat())
                    params.height = params.width
                    val margin = Util.dp2px(Constants.THUMB_MARGIN_DP.toFloat())
                    params.setMargins(margin, margin, margin, margin)
                    binding.offlineThumbnail.layoutParams = params
                    binding.offlineThumbnail.setImageURI(Uri.fromFile(file))
                } else {
                    binding.offlineThumbnail.setActualImageResource(
                        typeForName(
                            nodeOffline.name
                        ).iconResourceId
                    )
                }
            } else {
                binding.offlineThumbnail.setImageResource(
                    typeForName(
                        nodeOffline.name
                    ).iconResourceId
                )
            }
        } else {
            binding.offlineThumbnail.setImageResource(CoreUiR.drawable.ic_folder_list)
        }
        if (nodeOffline.isFolder && !Util.isOnline(requireContext())) {
            binding.optionShareLayout.visibility = View.GONE
            binding.separatorShare.visibility = View.GONE
        }
        super.onViewCreated(view, savedInstanceState)
    }

    override fun onDestroyView() {
        _binding = null
        super.onDestroyView()
    }

    override fun onClick(v: View) {
        when (v.id) {
            R.id.option_delete_offline_layout -> {
                if (parentFragment is OfflineNodeListener) {
                    (parentFragment as OfflineNodeListener).showConfirmationRemoveOfflineNode(
                        nodeOffline
                    )
                }
            }

            R.id.option_open_with_layout -> {
                OfflineUtils.openWithOffline(
                    requireContext(),
                    nodeOffline.handle.toLong()
                )
            }

            R.id.option_share_layout -> {
                OfflineUtils.shareOfflineNode(
                    requireContext(),
                    nodeOffline.handle.toLong()
                )
            }

            R.id.option_download_layout -> {
                (requireActivity() as ManagerActivity).saveOfflineNodesToDevice(listOf(nodeOffline))
            }

            R.id.option_properties_layout -> {
                val offlineIntent = Intent(requireContext(), OfflineFileInfoActivity::class.java)
                offlineIntent.putExtra(Constants.HANDLE, nodeOffline.handle)
                startActivity(offlineIntent)
            }
        }
        setStateBottomSheetBehaviorHidden()
    }

    companion object {
        private const val EXTRA_OFFLINE = "offline"

        fun newInstance(offline: MegaOffline) = OfflineOptionsBottomSheetDialogFragment().apply {
            arguments = bundleOf(EXTRA_OFFLINE to offline)
        }
    }
}